### ISSUE TITLE
Experiment with physical PV prediction using pvlib

### ISSUE THREAD

**Issue by JackKelly** [openclimatefix/power_perceiver#10]
- [x] find a PV system with shading issues
- [ ] predict PV power using:
    - [ ] pvlib fed with UKV NWP: See [this stack overflow question and answer](https://stackoverflow.com/questions/57067952/how-to-build-a-generic-pvsystem-using-pvlib). If that fails, also see: [`solarforecastarbiter.pvmodel.irradiance_to_power`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/solarforecastarbiter/pvmodel.py#L305-L341). See [this comment from Will](https://twitter.com/willholmgren/status/1497259792926076928).
        - [ ] maybe try inferring azimuth and angle using [`pvanalytics.system.infer_orientation_fit_pvwatts`](https://pvanalytics.readthedocs.io/en/stable/generated/pvanalytics.system.infer_orientation_fit_pvwatts.html#pvanalytics.system.infer_orientation_fit_pvwatts). Also see [this conversation with Will Holmgren on Twitter](https://twitter.com/willholmgren/status/1497258590721044480).
    - [ ] (pvlib fed with UKV NWP) -> an ML model which "corrects" pvlib
    - [ ] an ML model which is pre-trained on pvlib fed with UKV NWP
    - [ ] an ML model on its own
- [ ] compare performance of the above

# Context

I'm increasingly convinced that we need to accurately model each individual PV system as accurately as possible (especially local shading and inverter clipping). For two main reasons:

* When we train our ML model to infer PV power from satellite imagery, local effects like shading and inverter clipping are an unwelcome source of "noise".
* Likewise when we use PV power data from nearby PV systems to help predict PV power at the PV system of interest.

**Comment by JackKelly** [comment]
I've made a start on this on Friday, and found a few PV systems with pretty obvious shading issues (the plot shows power on the y axis and time-of-day on the x axis for June):

![image](https://user-images.githubusercontent.com/460756/155764351-e75cf1cf-4bb5-4d52-89e0-125776eb2bd6.png)

**Comment by JackKelly** [comment]
Here's what the pvlib prediction, using clearsky irradiance (and default wind speed and air temperature) looks like:

![image](https://user-images.githubusercontent.com/460756/156013080-c126c25a-adb0-46ad-86d4-b4d887a6598c.png)

Implemented in: https://github.com/openclimatefix/power_perceiver/blob/0f53822646da2605e4143268d6ac1be8cb4eabca/notebooks/2022-02-28_predict_pv_using_pvlib/predict_pv_using_pvlib.ipynb

Next step: Predict using NWPs

**Comment by JackKelly** [comment]
Actually, NWP irradiance isn't terrible:  (this is using the NWP irradiance produced by the most recent NWP model run).

![image](https://user-images.githubusercontent.com/460756/156195213-edd65c3b-8ee2-4040-a8ce-7522bc6e9ce2.png)

![image](https://user-images.githubusercontent.com/460756/156198822-111b3776-b944-431b-a56a-4acceba0a62f.png)

**Comment by JackKelly** [comment]
Hmm, I think one big source of error when using `pvlib` is likely to be the calculation of GHI, DNI and DHI.

The issue is that the NWP doesn't provide GHI, DNI and DHI. Instead it just provides "downwards short-wave radiation flux".

The [pvlib docs provide a method to compute GHI, DNI and DHI purely from NWP total cloud cover](https://pvlib-python.readthedocs.io/en/stable/forecasts.html#cloud-cover-and-radiation) (which already makes me nervous: we know that clouds at different vertical levels affect sunlight in different ways).

If we do this, then we get something that looks quite different to the NWP DSWRF:

![image](https://user-images.githubusercontent.com/460756/156226263-bca209c2-0b24-486c-91ac-f40089207642.png)

![image](https://user-images.githubusercontent.com/460756/156226286-4818b85a-2f3e-4dda-97bb-aa09b45c1ab6.png)

If we use NWP DSWRF as GHI (which I'd guess is the right thing to do?! But I'm really not sure! Especially because none of [the `pvlib.forecast` models](https://pvlib-python.readthedocs.io/en/stable/_modules/pvlib/forecast.html) do this) then we get this:

![image](https://user-images.githubusercontent.com/460756/156226397-0731adaa-3b58-4ec2-92ea-41b981ce0ca1.png)

**Comment by JackKelly** [comment]
[9408286](https://github.com/openclimatefix/power_perceiver/commit/9408286a5cabc736a5cca8531ad4b38c1b78dbba)

**Comment by JackKelly** [comment]
Next steps:

- [ ] Try predicting PV with DWSRF-as-GHI; vs GHI computed from cloud cover (directly above the point of interest).
- [ ] Try computing cloud cover using the NWP pixels in the Sun's line-of-sight (see issue #13).
- [ ] Try an ML model which sees the NWP pixels in the Sun's light-of-sight, plus a low res image of the pixels directly above the PV system (for light being scattered back).

**Comment by JackKelly** [comment]
MAE for predicting PV power with PVLib using NWP directly above the PV system (ignoring times when the sun is less than 10 degrees above the horizon):

* using DWSRF-as-GHI: 12.52% MAPE
* using GHI computed from cloud cover: 12.68% MAPE

Both these values are closer to 7% MAPE if using the complete time series (including nighttime).

**Comment by JackKelly** [comment]
to convert the vector line which represents the line-of-sight of the Sun through the lower 10 km of the atmosphere, I'm planning to adapt [Xiaolin Wu's line drawing algorithm](https://en.wikipedia.org/wiki/Xiaolin_Wu%27s_line_algorithm).

**Comment by JackKelly** [comment]
Give the model: solar azimuth, angle, and irradiance at the top of the atmosphere

**Comment by JackKelly** [comment]
Note to self: See the "next steps" in the ipython notebook.

**Comment by JackKelly** [comment]
Experiments computing GHI, DNI and DHI using the clouds, selecting the NWP grid boxes in the Sun's path:

* taking the max of each grid point for lcc, mcc, hcc.  total_cloud_cover is the max of that. 11.37% MAPE (only using times when elevation > 10). 6.84% when using nighttime too.
* taking the max of each grid point for lcc, mcc, hcc.  total_cloud_cover is the sum of that, clipped at 1. 11.19% MAPE (only using times when elevation > 10). 6.75% when using nighttime too.
*

**Comment by JackKelly** [comment]
Next steps:

- [ ] Speed up the code that gets NWP variables along the Sun's path. Load one timestep at a time from Zarr into memory, and then grab data from the data in memory.
- [ ] Train a boosted regression tree to predict PV. Try giving it PVLib's PV predictions. Or try without. The aim here is to get the best PV prediction from a model that see a single timestep. And compare ML model vs ML model fed with PVLib vs PVLib on its own.
- [ ] Then try training a boosted regression tree just to learn local factors of each PV system (i.e. don't tell it anything about clouds). (Ultimately, we'll probably train a neural net to do this, and to output a distribution. This should be useful to help downstream ML network "see" the effects of clouds).
- [ ] Then do the same with a deep neural net. Try training one PV system per neural net. Then try training one large neural net (giving it PV system ID) and see if it can still learn each PV system's characteristics.
